rpcs3/Emu/Cell/SPURecompiler.cpp

```diff
@@ -1904,6 +2070,97 @@ void spu_recompiler_base::old_interpreter(spu_thread& spu, void* ls, u8* /*rip*/
 	}
 }
 
+std::vector<u32> spu_thread::recover_functions(const void* ls_start, u32 /*entry*/)
+{
+	std::vector<u32> calls;
+	calls.reserve(100);
+
+	// Discover functions
+	// Use the most simple method: search for instructions that calls them
+	// And then filter invalid cases (does not detect tail calls)
+	for (u32 i = 0; i < SPU_LS_SIZE; i += 16)
+	{
+		// Search for BRSL and BRASL
+		// TODO: BISL
+		const v128 inst = read_from_ptr<be_t<v128>>(static_cast<const u8*>(ls_start), i);
+		const v128 shifted = gv_shr32(inst, 23);
+		const v128 eq_brsl = gv_eq32(shifted, v128::from32p(0x66));
+		const v128 eq_brasl = gv_eq32(shifted, v128::from32p(0x62));
+
+		if (!gv_testall0(eq_brsl))
+		{
+			if (eq_brsl.u32r[0])
+			{
+				calls.push_back(i + 0 * 4);
+			}
+
+			if (eq_brsl.u32r[1])
+			{
+				calls.push_back(i + 1 * 4);
+			}
+
+			if (eq_brsl.u32r[2])
+			{
+				calls.push_back(i + 2 * 4);
+			}
+
+			if (eq_brsl.u32r[3])
+			{
+				calls.push_back(i + 3 * 4);
+			}
+		}
+
+		if (!gv_testall0(eq_brasl))
+		{
+			if (eq_brasl.u32r[0])
+			{
+				calls.push_back(i + 0 * 4);
+			}
+
+			if (eq_brasl.u32r[1])
+			{
+				calls.push_back(i + 1 * 4);
+			}
+
+			if (eq_brasl.u32r[2])
+			{
+				calls.push_back(i + 2 * 4);
+			}
+
+			if (eq_brasl.u32r[3])
+			{
+				calls.push_back(i + 3 * 4);
+			}
```

Comment:
for (u32 i = 0; i < 4; i++) ?

Reply:
I think it produces inferior code but I guess it's not that bad because it is unlikely to pass the all-0 test first.